File: PCA/SVD.m
function [U,S,V]=SVD(A)

f_t = 0;
[m,n]=size(A);

if(m<n)
    A = A';
    [m,n]=size(A);
    f_t = 1;
end

[T,Va]=eig(A*A');
for i=1:m
    if(Va(i,i)<0)
        Va(i,i)=0;
    else
        Va(i,i)=sqrt(Va(i,i));
    end
end
[T1,~]=eig(A'*A);

for i=1:n
    Av=A*T1(:,n+1-i);
    US=T(:,m+1-i);
    if(Av(1,1)*US(1,1)<0)
        T(:,m+1-i)=-T(:,m+1-i);
    end
end

Va=Va(:,(m-n)+1:m);

if(f_t == 1)
    U=T1;
    S=Va';
    V=T;
else
    U=T;
    S=Va;
    V=T1;
end
% clc;
% clear;
% 
% f_t = 0;
% A = randn(10,5);
% [m,n]=size(A);
% 
% if(m<n)
%     A = A';
%     [m,n]=size(A);
%     f_t = 1;
% end
% 
% [T,V]=eig(A*A');
% for i=1:m
%     V(i,i)=sqrt(V(i,i));
% end
% [T1,V1]=eig(A'*A);
% 
% for i=1:n
%     Av=A*T1(:,n+1-i);
%     US=T(:,m+1-i);
%     if(Av(1,1)*US(1,1)<0)
%         T(:,m+1-i)=-T(:,m+1-i);
%     end
% end
% 
% V=V(:,(m-n)+1:m);
% 
% B=T*V*T1';
% 
% if(f_t == 1)
%     U11=T1;
%     S11=V';
%     V11=T;
%     A=A';
% else
%     U11=T;
%     S11=V;
%     V11=T1;
% end

File: PCA/SVD.py
import numpy
import numpy as np
import math


def SVD(A):
    f_t = 0
    m, n = A.shape
    # print(shape)
    if m < n:
        A = A.T
        f_t = 1
        m, n = A.shape
    S2, V = np.linalg.eig(np.dot(A.T, A))
    # print(V)
    for i in range(S2.size):
        S2[i] = math.sqrt(S2[i])
    index_value_S2 = sorted(enumerate(S2), key=lambda x: x[1], reverse=True)
    V_index = []
    S_value = []
    for index, value in index_value_S2:
        V_index.append(index)
        S_value.append(value)
    V = V[:, V_index]
    U = np.dot(A, V)

    zeros_matrix = np.zeros((m, m - n))
    U = np.c_[U, zeros_matrix]
    for index, value in enumerate(S_value):
        if value != 0:
            U[:, index] = U[:, index] / value
    S = np.diag(S_value)
    zeros_matrix = np.zeros((m-n, n))
    S = np.r_[S, zeros_matrix]
    # print(V_index)
    # print(S_value)
    # print(S)
    # # print(S2)
    # print(V)
    # print(U)
    print(U @ S @ V.T)
    return 0


A = np.array([[1, 0, 2], [2, 0, 4]])
matrix_normal = np.random.randn(5, 6)
print(matrix_normal)

SVD(matrix_normal)
